### ISSUE TITLE
differentiable top-k

### ISSUE THREAD

**Issue by lucidrains** [lucidrains/LVMAE-pytorch#2]
can't seem to find in the paper how they approached the differentiable top-k

**Comment by vasimr** [comment]
They could be referring to the method used in MoE, where you take the gate scores (post softmax) and truncate them to the top-k experts (effectively setting the other gates to 0). Then you multiply the output states by the new gates. 

```python
def top_k_route(router_logits, expert_outputs, K, is_training, noise_scale=1e-2):
  if is_training:
     router_logits+= noise_scale * randn_like(router_logits)
     
  probs = softmax(router_logits)
  gates, inds = probs.top_k(k=K)
  selection_states = expert_outputs.gather(inds)
  output = sum(gates*selection_states, dim=-2) # assuming (*,E,C)
  return output
```
This approach allows the gradient to flow through the gates to the top-k experts (tokens in this case), but has no gradient to the others. If you add noise to the logits prior to softmax, that will shuffle the selection set to ensure that you at least get some stochastic signal for the other high-ranking elements.

This can work quite well, even with K=1, but often requires a regularization loss to prevent collapse. Notably, the last time I used this, I also isolated the router with a stop-grad to avoid impacting the main model.

**Comment by lucidrains** [comment]
@vasimr yes, that could probably work well for MoE where you have 1 or 2 experts activated, but here K is a few orders larger

**Comment by lucidrains** [comment]
@vasimr may go with [this solution](https://github.com/lucidrains/CoLT5-attention/blob/main/colt5_attention/coor_descent.py) once i get around to this

**Comment by vasimr** [comment]
@lucidrains I'm not sure if that solution will work. First, it's iterative, which is not really conducive to training. And second, there seems to be a bug in it. Initializing `b = -s` => `sb_0=0` creates a selection bias, where if I were to use a random 2D score input, the resultant 1-hot scores become horizontal bands.

The method I described seems to work for large K (at least in a toy example).

```python
def top_k_select(logits, k, is_training=False, noise_scale=1e-2):
    if is_training:
        logits+= noise_scale * torch.randn_like(logits)

    B, *dims = logits.shape
    logits = logits.view(B, -1)
    probs = torch.softmax(logits, dim=-1)
    gates, indices = torch.topk(probs, k, dim=-1, largest=True, sorted=False)
    values = torch.zeros_like(router_logits)
    scores = scores.scatter_(-1, indices, values)
    
    return scores.view(B, *dims)
```

This correctly selects a subset (e.g. logits of size (1,32,32) and K=256), while retaining gradients to the selected top-k. However, the resultant scores are mostly near 0.0 for the distribution I tried (skewed normal), so normalization by the max score (along the batch dim) may help if used as a weighting term. Notably, the max score in this case becomes a dynamic constant, but does not require back-prop.

**Comment by lucidrains** [comment]
@vasimr i have a fused version [here](https://github.com/lucidrains/CoLT5-attention/blob/main/colt5_attention/triton_coor_descent.py).

will take a look at the init, but last i checked, the repo is functional and i'm even able to train working autoregressive language models with differentiable topk in place of softmax in attention.

**Comment by lucidrains** [comment]
@vasimr i'm not sure your way would work, given skewed distribution
